commit 789a37f14405e2d1a05a76c9fb4ed2d49d4580d5
Author: guoyiyuan <yguoaz@gmail.com>
Date:   Wed Jul 13 20:55:51 2022 +0800

    Prevent potential buffer overflow for large value of php_cli_server_workers_max
    
    Fixes #8989.
    Closes #9000.

diff --git a/NEWS b/NEWS
index fdf9cb3c93..682404e406 100644
--- a/NEWS
+++ b/NEWS
@@ -2,6 +2,10 @@ PHP                                                                        NEWS
 |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
 ?? ??? 2022, PHP 8.0.22
 
+- CLI:
+  . Fixed potential overflow for the builtin server via the PHP_CLI_SERVER_WORKERS
+    environment variable. (yiyuaner)
+
 - Core:
   . Fixed bug GH-8923 (error_log on Windows can hold the file write lock). (cmb)
 
diff --git a/sapi/cli/php_cli_server.c b/sapi/cli/php_cli_server.c
index 726eee55c1..a1fc2cf0f8 100644
--- a/sapi/cli/php_cli_server.c
+++ b/sapi/cli/php_cli_server.c
@@ -2299,7 +2299,7 @@ static void php_cli_server_dtor(php_cli_server *server) /* {{{ */
 					  !WIFSIGNALED(php_cli_server_worker_status));
 		}
 
-		free(php_cli_server_workers);
+		pefree(php_cli_server_workers, 1);
 	}
 #endif
 } /* }}} */
@@ -2385,12 +2385,8 @@ static void php_cli_server_startup_workers() {
 	if (php_cli_server_workers_max > 1) {
 		zend_long php_cli_server_worker;
 
-		php_cli_server_workers = calloc(
-			php_cli_server_workers_max, sizeof(pid_t));
-		if (!php_cli_server_workers) {
-			php_cli_server_workers_max = 1;
-			return;
-		}
+		php_cli_server_workers = pecalloc(
+			php_cli_server_workers_max, sizeof(pid_t), 1);
 
 		php_cli_server_master = getpid();
 
